Do not download initial headers from inbound peers

ptschip opened the pull request:
When we start syncing and we are still in IsInitialBlockdownload()
then do not request the initial batches of headers from inbound peers.
They are the least trustworthy peers, so we should rather only download
the headers from outbound peers, until we've completed initial sync.

ptschip committed:
Do not download initial headers from inbound peers

When we start syncing and we are still in IsInitialBlockdownload()
then do not request the initial batches of headers from inbound peers.
They are the least trustworthy peers, so we should rather only download
the headers from outbound peers, until we've completed initial sync.

sickpig commented:
we are not able to get why travis is failing cause the travis log got bigger than 4MB which is an hard limit hence travis cut it w/o revealing the actual problem

sickpig commented:
actually I was able to find the coverage report of the test buried in the log: 

```
TEST                           | PASSED | DURATION
cashlibtest.py                 | True   | 0 s
tweak.py                       | True   | 1 s
miningtest.py                  | True   | 5 s
may152018_forkactivation_2.py  | True   | 4 s
grapheneblocks.py              | True   | 13 s
notify.py                      | True   | 12 s
may152018_forkactivation_1.py  | True   | 12 s
validateblocktemplate.py       | True   | 37 s
wallet.py                      | True   | 47 s
wallet-dump.py                 | True   | 5 s
bip68-112-113-p2p.py           | True   | 58 s
listtransactions.py            | True   | 17 s
wallet-hd.py                   | True   | 42 s
mempool_resurrect_test.py      | True   | 2 s
receivedby.py                  | True   | 12 s
txn_doublespend.py --mineblock | True   | 12 s
txn_clone.py                   | True   | 10 s
getchaintips.py                | True   | 11 s
rawtransactions.py             | True   | 21 s
rest.py                        | True   | 14 s
mempool_spendcoinbase.py       | True   | 4 s
mempool_limit.py               | True   | 6 s
httpbasics.py                  | True   | 2 s
mempool_reorg.py               | True   | 11 s
multi_rpc.py                   | True   | 3 s
proxy_test.py                  | True   | 2 s
zapwallettxes.py               | True   | 13 s
merkle_blocks.py               | True   | 14 s
signrawtransactions.py         | True   | 1 s
fundrawtransaction.py          | True   | 34 s
nodehandling.py                | True   | 11 s
walletbackup.py                | True   | 40 s
decodescript.py                | True   | 2 s
reindex.py                     | True   | 12 s
disablewallet.py               | True   | 0 s
blockchain.py                  | True   | 16 s
keypool.py                     | True   | 4 s
sendheaders.py                 | True   | 26 s
invalidtxrequest.py            | True   | 24 s
abandonconflict.py             | True   | 12 s
excessive.py                   | True   | 193 s
importprunedfunds.py           | True   | 6 s
zmq_test.py                    | True   | 8 s
p2p-versionbits-warning.py     | True   | 21 s
thinblocks.py                  | True   | 14 s
parallel.py                    | False  | 1202 s

ALL                            | False  | 2016 s (accumulated)
```
so the failing test is `parallel.py`

awemany commented:
@sickpig: This is because I added the dump of all the `debug.log`s which will overflow this sometimes. But to me it looks like this was a timeout? See line 4920.

ptschip committed:
Remove the check for inbound peers when syncing

while this works on mainnet it only add yet another regest headache
which requires yet another networkid workaround for testing. Furthermore
since this fix is only a partial fix in itself we should not use it
and rather go for the more comprehensive fix

ptschip committed:
AskFor() blocks from all peers where chain height is < 1024

This solves an initial sync issue whereby a peer connects to us inbound
and gives us headers but no blocks, before any other peers can connect
or be connected to. This results in the request manager not being able
to re-request those blocks when they timeout because it doesn't yet
know abou them.

ptschip commented:
I think this is a more comprehensive solution.  The first solution was really a band aid and then was impacting our regression tests.

I've tested the current solution on mainnet and in every case it was able to recover within about 1 minute or so of the beginning of sync.

ptschip commented:
closing in favor of #1252
